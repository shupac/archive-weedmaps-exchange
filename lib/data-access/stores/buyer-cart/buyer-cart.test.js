import { getSnapshot } from 'mobx-state-tree';
import mockCart, {
  buyerCartItem,
  mockErrorCart,
  mockCartErrors,
} from 'lib/mocks/cart';
import mockCartOrder from 'lib/mocks/cart-order';
import BuyerCart from './';

function setup() {
  const mockFetchClient = {
    post: jest.fn().mockReturnValue(Promise.resolve({ data: mockCart })),
    patch: jest.fn().mockReturnValue(Promise.resolve({ data: mockCart })),
    fetch: jest.fn().mockReturnValue(Promise.resolve({ data: mockCart })),
  };

  const buyerCart = BuyerCart.create(
    { cart: mockCart },
    {
      client: mockFetchClient,
    },
  );
  return { buyerCart, mockFetchClient };
}

const cartItems = [
  { variant_id: 'e0c6412f-05e4-4779-b964-86932d4de96a', quantity: 2 },
];
const updatedCartItems = {
  id: 'e0c6412f-05e4-4779-b964-86932d4de96a',
  quantity: 3,
};

describe('Buyer Cart Store', () => {
  it('can fetch buyer from cart endpoint and set data', async () => {
    const { buyerCart } = setup();
    await buyerCart.fetchCart();
    expect(getSnapshot(buyerCart.cart)).toMatchSnapshot();
  });

  it('will compute the cart item count', () => {
    const { buyerCart } = setup();
    expect(buyerCart.cartItemCount).toBe(2);
  });

  it('will compute the cart item by brand', () => {
    const { buyerCart } = setup();
    expect(buyerCart.cartItemsByBrand).toEqual(buyerCartItem);
  });

  it('will compute the cart errors', () => {
    const buyerCart = BuyerCart.create({ cart: mockErrorCart });
    expect(buyerCart.cartErrors).toEqual(mockCartErrors);
  });

  describe('When adding items to cart', () => {
    it('will post to cart endpoint ', () => {
      const { buyerCart, mockFetchClient } = setup();

      buyerCart.addCartItems(cartItems);
      expect(mockFetchClient.post).toHaveBeenCalledWith('/buyer/cart/items', {
        data: [
          { quantity: 2, variant_id: 'e0c6412f-05e4-4779-b964-86932d4de96a' },
        ],
      });
    });

    it('will handle errors ', () => {
      const mockErrorClient = {
        post: () => {
          throw new Error('Test');
        },
        patch: () => {
          throw new Error('Test');
        },
        fetch: () => {
          throw new Error('Test');
        },
      };
      const buyerCart = BuyerCart.create(
        { cart: mockCart },
        {
          client: mockErrorClient,
        },
      );
      const log = jest.spyOn(global.console, 'error').mockReturnValue();
      buyerCart.addCartItems(cartItems);
      expect(log).toHaveBeenCalledWith(new Error('Test'));
      buyerCart.updateCartItem(updatedCartItems);
      expect(log).toHaveBeenCalledWith(new Error('Test'));
      buyerCart.fetchCart(cartItems);
      expect(log).toHaveBeenCalledWith(new Error('Test'));
    });
  });
  describe('When updating items to cart', () => {
    it('will post to cart endpoint ', () => {
      const { buyerCart, mockFetchClient } = setup();

      buyerCart.updateCartItem(updatedCartItems);
      expect(mockFetchClient.patch).toHaveBeenCalledWith(
        '/buyer/cart/items/e0c6412f-05e4-4779-b964-86932d4de96a',
        {
          data: {
            attributes: { quantity: 3 },
            id: 'e0c6412f-05e4-4779-b964-86932d4de96a',
            type: 'cart_item',
          },
        },
      );
    });
  });
  describe('When checkout cart', () => {
    it('will post to cart endpoint when checking out cart', () => {
      const { buyerCart, mockFetchClient } = setup();

      buyerCart.checkoutCart();
      expect(mockFetchClient.post).toHaveBeenCalledWith('/shopping/checkout');
    });
  });

  describe('Fetch cart order', () => {
    it('will fetch cart order', () => {
      const mockFetchClient = {
        fetch: jest
          .fn()
          .mockReturnValue(Promise.resolve({ data: mockCartOrder })),
      };
      const buyerCart = BuyerCart.create(
        { cartOrderData: mockCartOrder },
        {
          client: mockFetchClient,
        },
      );
      buyerCart.fetchCartOrder(mockCartOrder.id);
      expect(mockFetchClient.fetch).toHaveBeenCalledWith(
        `/shopping/orders/${mockCartOrder.id}`,
      );
    });

    it('will calculate the total values for the order', () => {
      const mockFetchClient = {
        fetch: jest
          .fn()
          .mockReturnValue(Promise.resolve({ data: mockCartOrder })),
      };
      const buyerCart = BuyerCart.create(
        { cartOrderData: mockCartOrder },
        {
          client: mockFetchClient,
        },
      );

      expect(buyerCart.cartOrder.subtotal).toEqual(194250);
      expect(buyerCart.cartOrder.shippingFee).toEqual(200);
      expect(buyerCart.cartOrder.total).toEqual(194450);
      expect(buyerCart.cartOrder.totalItems).toEqual(8);
    });

    it('will handle errors', () => {
      const mockErrorClient = {
        post: () => {
          throw new Error('Test');
        },
        fetch: () => {
          throw new Error('Test');
        },
      };
      const buyerCart = BuyerCart.create(
        { cart: mockCart },
        {
          client: mockErrorClient,
        },
      );
      const log = jest.spyOn(global.console, 'error').mockReturnValue();
      buyerCart.fetchCartOrder('foo');
      expect(log).toHaveBeenCalledWith(new Error('Test'));
      log.mockRestore();
    });
  });
});
