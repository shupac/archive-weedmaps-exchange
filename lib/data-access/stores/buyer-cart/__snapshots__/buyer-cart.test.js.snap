// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Buyer Cart Store can fetch buyer from cart endpoint and set data 1`] = `
Object {
  "brands": Array [
    Object {
      "deliveryEta": Object {
        "etaMax": 3,
        "etaMaxUnit": "day",
        "etaMin": 1,
        "etaMinUnit": "day",
      },
      "description": "this is the brand description",
      "id": "d054b34f-428e-4583-a2a4-27f4e5aea6f3",
      "minimumPurchasePrice": 0,
      "name": "GfarmaLabs",
      "shippingFee": 0,
      "slug": "gfarmalabs",
    },
  ],
  "cartErrors": Array [],
  "id": "238ee8c0-0297-44c0-b768-41397260df40",
  "items": Array [
    Object {
      "amount": 3,
      "id": "126974e5-a216-4b63-98c8-d065c0180125",
      "price": 2800,
      "variantId": "630580e0-c20b-48fc-8ea1-e03abe831b05",
    },
    Object {
      "amount": 3,
      "id": "6b1e92e0-ce45-442a-b54d-e6e9ba617614",
      "price": 300,
      "variantId": "67879046-6491-4f6b-869b-75e94adeb9f0",
    },
  ],
  "shippingFee": 0,
  "subtotal": 93890,
  "total": 93890,
}
`;
